### ISSUE TITLE
"with engine.begin()"-style execution failure with mysqldb but not with pymysql

### ISSUE THREAD

**Issue by gordthompson** [sqlalchemy/sqlalchemy#5154]
Discovered this in test failures for

https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1721/

When using server-side cursors, the old `engine.execute` works for both `pymysql` and `mysqldb`. Changing to the now-preferred `with engine.begin() as conn:` method still works for `pymysql`, but for `mysqldb` it results in 

> MySQLdb._exceptions.ProgrammingError: (2014, "Commands out of sync; you can't run this command now")

Repo code:

```python
import sqlalchemy as sa

engine_pymysql = sa.create_engine(
    "mysql+pymysql://root:toot@localhost:3307/mydb", echo=True
)
engine_mysqldb = sa.create_engine(
    "mysql+mysqldb://root:toot@localhost:3307/mydb", echo=True
)

stmt = sa.select([1]).with_for_update().execution_options(stream_results=True)

print(">>> engine.execute - pymysql")
result = engine_pymysql.execute(stmt)
print(">>> ...OK")
print(">>> engine.execute - mysqldb")
result = engine_mysqldb.execute(stmt)
print(">>> ...OK")
print(">>> context handler - pymysql")
with engine_pymysql.begin() as conn:
    result = conn.execute(stmt)
print(">>> ...OK")
print("context handler - mysqldb")
with engine_mysqldb.begin() as conn:
    result = conn.execute(stmt)
print(">>> ...OK")
```

Console output:

```
>>> engine.execute - pymysql
2020-02-13 16:30:43,016 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'sql_mode'
2020-02-13 16:30:43,016 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,018 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'lower_case_table_names'
2020-02-13 16:30:43,018 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,020 INFO sqlalchemy.engine.base.Engine SELECT DATABASE()
2020-02-13 16:30:43,020 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,021 INFO sqlalchemy.engine.base.Engine show collation where `Charset` = 'utf8mb4' and `Collation` = 'utf8mb4_bin'
2020-02-13 16:30:43,021 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,024 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,024 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,025 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,025 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,026 INFO sqlalchemy.engine.base.Engine SELECT CAST('test collated returns' AS CHAR CHARACTER SET utf8mb4) COLLATE utf8mb4_bin AS anon_1
2020-02-13 16:30:43,026 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,028 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,029 INFO sqlalchemy.engine.base.Engine {}
>>> ...OK
>>> engine.execute - mysqldb
2020-02-13 16:30:43,037 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'sql_mode'
2020-02-13 16:30:43,037 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,039 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'lower_case_table_names'
2020-02-13 16:30:43,039 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,041 INFO sqlalchemy.engine.base.Engine SELECT DATABASE()
2020-02-13 16:30:43,041 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,042 INFO sqlalchemy.engine.base.Engine show collation where `Charset` = 'utf8mb4' and `Collation` = 'utf8mb4_bin'
2020-02-13 16:30:43,042 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,044 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,044 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,045 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,045 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,046 INFO sqlalchemy.engine.base.Engine SELECT CAST('test collated returns' AS CHAR CHARACTER SET utf8mb4) COLLATE utf8mb4_bin AS anon_1
2020-02-13 16:30:43,047 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,048 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,048 INFO sqlalchemy.engine.base.Engine ()
>>> ...OK
>>> context handler - pymysql
2020-02-13 16:30:43,049 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-02-13 16:30:43,050 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,050 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,052 INFO sqlalchemy.engine.base.Engine COMMIT
>>> ...OK
context handler - mysqldb
2020-02-13 16:30:43,054 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-02-13 16:30:43,055 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,055 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,055 INFO sqlalchemy.engine.base.Engine COMMIT
C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\pymysql\connections.py:756: UserWarning: Previous unbuffered result was left incomplete
  warnings.warn("Previous unbuffered result was left incomplete")
Traceback (most recent call last):
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 755, in _commit_impl
    self.engine.dialect.do_commit(self.connection)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\dialects\mysql\base.py", line 2290, in do_commit
    dbapi_connection.commit()
MySQLdb._exceptions.ProgrammingError: (2014, "Commands out of sync; you can't run this command now")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:/Users/Gord/PycharmProjects/sqla_mysql/main.py", line 24, in <module>
    result = conn.execute(stmt)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 2064, in __exit__
    self.transaction.commit()
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1727, in commit
    self._do_commit()
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1758, in _do_commit
    self.connection._commit_impl()
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 757, in _commit_impl
    self._handle_dbapi_exception(e, None, None, None, None)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1476, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 755, in _commit_impl
    self.engine.dialect.do_commit(self.connection)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\dialects\mysql\base.py", line 2290, in do_commit
    dbapi_connection.commit()
sqlalchemy.exc.ProgrammingError: (MySQLdb._exceptions.ProgrammingError) (2014, "Commands out of sync; you can't run this command now")
(Background on this error at: http://sqlalche.me/e/f405)

Process finished with exit code 1
```

**Comment by zzzeek** [comment]
pymysql is emitting a warning also; make sure the results from the SELECT statement are fully fetched, e.g. result.fetchall() or result.close()



On Thu, Feb 13, 2020, at 6:46 PM, Gord Thompson wrote:
> Discovered this in test failures for

> https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1721/

> When using server-side cursors, the old `engine.execute` works for both `pymysql` and `mysqldb`. Changing to the now-preferred `with engine.begin() as conn:` method still works for `pymysql`, but for `mysqldb` it results in

>> MySQLdb._exceptions.ProgrammingError: (2014, "Commands out of sync; you can't run this command now")

> Repo code:

> import sqlalchemy as sa

engine_pymysql = sa.create_engine(
    "mysql+pymysql://root:toot@localhost:3307/mydb", echo=True
)
engine_mysqldb = sa.create_engine(
    "mysql+mysqldb://root:toot@localhost:3307/mydb", echo=True
)

stmt = sa.select([1]).with_for_update().execution_options(stream_results=True)

> print(">>> engine.execute - pymysql")
result = engine_pymysql.execute(stmt)
> print(">>> ...OK")
> print(">>> engine.execute - mysqldb")
result = engine_mysqldb.execute(stmt)
> print(">>> ...OK")
> print(">>> context handler - pymysql")
> with engine_pymysql.begin() as conn:
    result = conn.execute(stmt)
> print(">>> ...OK")
> print("context handler - mysqldb")
> with engine_mysqldb.begin() as conn:
    result = conn.execute(stmt)
> print(">>> ...OK")
> Console output:

> `>>> engine.execute - pymysql
2020-02-13 16:30:43,016 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'sql_mode'
2020-02-13 16:30:43,016 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,018 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'lower_case_table_names'
2020-02-13 16:30:43,018 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,020 INFO sqlalchemy.engine.base.Engine SELECT DATABASE()
2020-02-13 16:30:43,020 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,021 INFO sqlalchemy.engine.base.Engine show collation where `Charset` = 'utf8mb4' and `Collation` = 'utf8mb4_bin'
2020-02-13 16:30:43,021 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,024 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,024 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,025 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,025 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,026 INFO sqlalchemy.engine.base.Engine SELECT CAST('test collated returns' AS CHAR CHARACTER SET utf8mb4) COLLATE utf8mb4_bin AS anon_1
2020-02-13 16:30:43,026 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,028 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,029 INFO sqlalchemy.engine.base.Engine {}
>>> ...OK
>>> engine.execute - mysqldb
2020-02-13 16:30:43,037 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'sql_mode'
2020-02-13 16:30:43,037 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,039 INFO sqlalchemy.engine.base.Engine SHOW VARIABLES LIKE 'lower_case_table_names'
2020-02-13 16:30:43,039 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,041 INFO sqlalchemy.engine.base.Engine SELECT DATABASE()
2020-02-13 16:30:43,041 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,042 INFO sqlalchemy.engine.base.Engine show collation where `Charset` = 'utf8mb4' and `Collation` = 'utf8mb4_bin'
2020-02-13 16:30:43,042 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,044 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,044 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,045 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS CHAR(60)) AS anon_1
2020-02-13 16:30:43,045 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,046 INFO sqlalchemy.engine.base.Engine SELECT CAST('test collated returns' AS CHAR CHARACTER SET utf8mb4) COLLATE utf8mb4_bin AS anon_1
2020-02-13 16:30:43,047 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,048 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,048 INFO sqlalchemy.engine.base.Engine ()
>>> ...OK
>>> context handler - pymysql
2020-02-13 16:30:43,049 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-02-13 16:30:43,050 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,050 INFO sqlalchemy.engine.base.Engine {}
2020-02-13 16:30:43,052 INFO sqlalchemy.engine.base.Engine COMMIT
>>> ...OK
context handler - mysqldb
2020-02-13 16:30:43,054 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-02-13 16:30:43,055 INFO sqlalchemy.engine.base.Engine SELECT 1 FOR UPDATE
2020-02-13 16:30:43,055 INFO sqlalchemy.engine.base.Engine ()
2020-02-13 16:30:43,055 INFO sqlalchemy.engine.base.Engine COMMIT
C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\pymysql\connections.py:756: UserWarning: Previous unbuffered result was left incomplete
  warnings.warn("Previous unbuffered result was left incomplete")
Traceback (most recent call last):
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 755, in _commit_impl
    self.engine.dialect.do_commit(self.connection)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\dialects\mysql\base.py", line 2290, in do_commit
    dbapi_connection.commit()
MySQLdb._exceptions.ProgrammingError: (2014, "Commands out of sync; you can't run this command now")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:/Users/Gord/PycharmProjects/sqla_mysql/main.py", line 24, in <module>
    result = conn.execute(stmt)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 2064, in __exit__
    self.transaction.commit()
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1727, in commit
    self._do_commit()
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1758, in _do_commit
    self.connection._commit_impl()
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 757, in _commit_impl
    self._handle_dbapi_exception(e, None, None, None, None)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1476, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\engine\base.py", line 755, in _commit_impl
    self.engine.dialect.do_commit(self.connection)
  File "C:\Users\Gord\PycharmProjects\sqla_mysql\venv\lib\site-packages\sqlalchemy\dialects\mysql\base.py", line 2290, in do_commit
    dbapi_connection.commit()
sqlalchemy.exc.ProgrammingError: (MySQLdb._exceptions.ProgrammingError) (2014, "Commands out of sync; you can't run this command now")
(Background on this error at: http://sqlalche.me/e/f405)

Process finished with exit code 1
`

**Comment by zzzeek** [comment]
this worked right we can close ?

**Comment by gordthompson** [comment]
I asked about this [on Gerrit](https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1723/) over the weekend:

> Should I add "Fixes: #5154" to the commit message to link this with my GitHub question?

**Comment by zzzeek** [comment]
This is just within the test suite refactor so without a top-level issue to track that (also not critical) we can just close this.
